File: backend/app/schemas/lease.py
from pydantic import BaseModel
from datetime import datetime, date
from typing import Optional
from uuid import UUID
from decimal import Decimal


class LeaseBase(BaseModel):
    property_id: UUID
    tenant_id: UUID
    organization_id: UUID
    start_date: date
    end_date: Optional[date] = None
    monthly_rent: Decimal
    charges: Optional[Decimal] = None
    deposit: Optional[Decimal] = None
    indexation_rate: Optional[Decimal] = None


class LeaseCreate(LeaseBase):
    pass


class LeaseUpdate(BaseModel):
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    monthly_rent: Optional[Decimal] = None
    charges: Optional[Decimal] = None
    deposit: Optional[Decimal] = None
    indexation_rate: Optional[Decimal] = None


class Lease(LeaseBase):
    id: UUID
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: lib/validations/lease.ts
import { z } from "zod";

export const leaseSchema = z.object({
  property_id: z.string().uuid("ID propriété invalide"),
  tenant_id: z.string().uuid("ID locataire invalide"),
  start_date: z.string().min(1, "Date de début requise"),
  end_date: z.string().min(1, "Date de fin requise"),
  monthly_rent: z.number().min(0, "Le loyer doit être positif"),
  charges: z.number().min(0, "Les charges doivent être positives"),
  deposit: z.number().min(0, "Le dépôt doit être positif"),
  indexation_clause: z.boolean().default(false),
  indexation_rate: z.number().optional(),
  payment_day: z.number().min(1).max(31).default(1),
  termination_notice_period: z.number().optional(),
  renewal_conditions: z.string().optional(),
  special_clauses: z.array(z.string()).optional(),
}).refine(
  (data) => new Date(data.end_date) > new Date(data.start_date),
  {
    message: "La date de fin doit être après la date de début",
    path: ["end_date"],
  }
);

export const updateLeaseSchema = leaseSchema.partial().extend({
  status: z.enum(["active", "expired", "terminated", "pending"]).optional(),
});

export type LeaseFormData = z.infer<typeof leaseSchema>;
export type UpdateLeaseFormData = z.infer<typeof updateLeaseSchema>;
